blob/service.go

```diff
@@ -49,26 +50,131 @@ type Submitter interface {
 }
 
 type Service struct {
+	// ctx represents the Service's lifecycle context.
+	ctx    context.Context
+	cancel context.CancelFunc
 	// accessor dials the given celestia-core endpoint to submit blobs.
 	blobSubmitter Submitter
 	// shareGetter retrieves the EDS to fetch all shares from the requested header.
 	shareGetter share.Getter
 	// headerGetter fetches header by the provided height
 	headerGetter func(context.Context, uint64) (*header.ExtendedHeader, error)
+	// headerSub subscribes to new headers to supply to blob subscriptions.
+	headerSub func(ctx context.Context) (<-chan *header.ExtendedHeader, error)
+
+	// activeSubscriptions tracks the number of active subscriptions
+	activeSubscriptions sync.WaitGroup
+	stopped             bool
+	mu                  sync.Mutex
 }
 
 func NewService(
 	submitter Submitter,
 	getter share.Getter,
 	headerGetter func(context.Context, uint64) (*header.ExtendedHeader, error),
+	headerSub func(ctx context.Context) (<-chan *header.ExtendedHeader, error),
 ) *Service {
 	return &Service{
 		blobSubmitter: submitter,
 		shareGetter:   getter,
 		headerGetter:  headerGetter,
+		headerSub:     headerSub,
 	}
 }
 
+func (s *Service) Start(context.Context) error {
+	s.ctx, s.cancel = context.WithCancel(context.Background())
+	return nil
+}
+
+func (s *Service) Stop(context.Context) error {
+	s.mu.Lock()
+	if s.cancel != nil {
+		s.stopped = true
+		s.cancel()
+	}
+	s.mu.Unlock()
+
+	s.activeSubscriptions.Wait()
+	return nil
+}
+
+type SubscriptionResponse struct {
+	Blobs  []*Blob
+	Height uint64
+	Error  error
+}
+
+func (s *Service) Subscribe(ctx context.Context, ns share.Namespace) (<-chan *SubscriptionResponse, error) {
+	s.mu.Lock()
+	if s.stopped {
+		s.mu.Unlock()
+		return nil, fmt.Errorf("service has been stopped")
+	}
+	if s.ctx == nil {
+		s.mu.Unlock()
+		return nil, fmt.Errorf("service has not been started")
+	}
+	s.activeSubscriptions.Add(1)
+	s.mu.Unlock()
+
+	headerCh, err := s.headerSub(ctx)
+	if err != nil {
+		s.activeSubscriptions.Done()
+		return nil, err
+	}
+
+	blobCh := make(chan *SubscriptionResponse, 16)
+	go func() {
+		defer s.activeSubscriptions.Done()
+		defer close(blobCh)
+
+		ticker := time.NewTicker(time.Minute)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case header, ok := <-headerCh:
+				if ctx.Err() != nil {
+					log.Debug("blobsub: canceling subscription due to user context closing")
+					return
+				}
+				if !ok {
+					log.Errorw("header channel closed for subscription", "namespace", ns.ID())
+					return
+				}
+				if len(blobCh) == cap(blobCh)-1 {
+					log.Warn("blobsub: slow consumer detected, closing subscription")
+					// still send an error to the client even though it seems like they've stopped consuming messages
+					blobCh <- &SubscriptionResponse{Error: errors.New("canceling subscription due to slow consumption")}
+					return
+				}
```

Comment:
There is no reason to send message to the channel, because it will be closed immediately after return and the messages will be gone.
 I would suggest to simplify it a bit by adding default case to the select where you write to blobCh.
```
default:
log.Warn("blobsub: slow consumer detected, closing subscription")
return
```

Reply:
Buffered messages can still be read from a closed channel no?

Reply:
True

Reply:
Literally was pending comment
![image](https://github.com/user-attachments/assets/8baed259-db4b-4db1-9f86-23055f8972f5)